docs/runtime/builtin-base-methods.md

```diff
@@ -0,0 +1,96 @@
+be exported to the Base library. This DSL facilitates automatic parsing
+and typechecking of method arguments and wrapping the node in an actual
+`Function` object.
+
+## `@BuiltinMethod` Annotation
+The `@BuiltinMethod` annotation is applied to a Truffle Node subclass that
+should be treated as a builtin method. It takes the following arguments:
+
+1. `type` **String** the name of the type the method will be registered on.
+2. `name` **String** the method name.
+3. `description` **String** a summary of the method's behavior, for
+   documentation purposes.
+4. `alwaysDirect` **Boolean = `true`** describes whether the function can always
+   be safely called directly. It should be set to `false` for methods that
+   evaluate user code in a tail position (e.g. `if_then_else`). If the method
+   does not take functions or thunks as parameters, or it only evaluates these
+   in a non-tail position, this parameter should be left defaulted.
+
+The annotation will generate a `RootNode` subclass in the same package as the
+declaring node, with a name basing on the declaring node's name. The name of
+the generated node is the name of the original node, with the `Node` suffix
+stripped and the `MethodGen` suffix appended. E.g. `AndOperatorNode` will
+generate a `AndOperatorMethodGen` root node.
+
+## Structural Requirements
+The DSL places certain structural requirements on the declaring node.
+
+1. Construction. The node must be constructible using either a paremeterless,
+   static `build` method or a parameterless constructor. If a suitable `build`
+   method is defined, it will be preferred over the constructor.
+2. Execution. The node must declare a single, accessible
+   (i.e. at least package-private), non-void method named `execute`.
+   
+## `Execute` Method Semantics
+The `execute` method defined by the declaring node has the following semantics.
+
+1. Return type. If the return type is `Stateful`, the method is assumed to
+   modify the monadic state. For any other return type, the method cannot
+   modify the monadic state.
+2. Arguments. The method may take any number of arguments, though at least
+   an argument named `self` must be present. The arguments have following
+   semantics:
+   1. An `Object` argument may be annotated with `@MonadicState`. This tells
+      the DSL to substitute it for the current value of monadic state. Note
+      that this value is read-only. In order to modify state, the method
+      must return a `Stateful`.
+   2. Any `VirtualFrame` argument will be passed the current execution frame.
+   3. Any `CallerInfo` argument will be passed the current `CallerInfo` and
+      mark the method as requiring the caller info at runtime.
+   4. All other arguments are treated as positional arguments to the method.
+      Note that according to the language specification, any method must take
+      an argument named `this`. Because of a naming conflict with Java, this
```

Comment:
This naming conflict is pretty confusing. Could we not do something like `ensoThis`, given we do call it `this` in Enso?